ballerina/modules/wssec/ws_security_methods.bal

```diff
@@ -0,0 +1,204 @@
+// under the License.
+import ballerina/crypto;
+import ballerina/regex;
+
+xmlns "http://schemas.xmlsoap.org/soap/envelope/" as soap;
+xmlns "http://www.w3.org/2000/09/xmldsig#" as ds;
+
+function addSecurityHeader(Document document) returns WSSecurityHeader|Error {
+    WSSecurityHeader wsSecHeader = check new (document);
+    Error? insertHeader = wsSecHeader.insertSecHeader();
+    if insertHeader is () {
+        return wsSecHeader;
+    }
+    return insertHeader;
+}
+
+public function decryptData(byte[] cipherText, EncryptionAlgorithm encryptionAlgorithm,
+                            crypto:PublicKey|crypto:PrivateKey key) returns byte[]|Error {
+    Encryption encrypt = check new ();
+    return encrypt.decryptData(cipherText, encryptionAlgorithm, key);
+}
+
+public function verifyData(byte[] data, byte[] signature, crypto:PublicKey publicKey,
+                           SignatureAlgorithm signatureAlgorithm) returns boolean|Error {
+    Signature sign = check new ();
+    return sign.verifySignature(data, signature, publicKey, signatureAlgorithm);
+}
+
+function addSignature(Signature sign, string signatureAlgorithm, byte[] signature) returns Signature|Error {
+    sign.setSignatureAlgorithm(signatureAlgorithm);
+    sign.setSignatureValue(signature);
+    return sign;
+}
+
+function addEncryption(Encryption encrypt, string encryptionAlgorithm, byte[] encryption) returns Encryption|Error {
+    encrypt.setEncryptionAlgorithm(encryptionAlgorithm);
+    encrypt.setEncryptedData(encryption);
+    return encrypt;
+}
+
+function applyEncryptedKey(string envelopeString, crypto:PrivateKey symmetricKey, crypto:PublicKey encryptKey)
+    returns string|Error {
+    string securedEnvelope = envelopeString;
+    do {
+        Encryption encryption = check new ();
+        byte[] encryptedKey = check crypto:encryptRsaEcb(symmetricKey.toBalString().toBytes(), encryptKey);
+        string encryptedKeyElements = check encryption.getEncryptedKeyElements(encryptedKey);
+        string replace = regex:replace(encryptedKeyElements, string `<?.*?><`, "<");
+        string:RegExp securityToken = 
+            re `<wsse:SecurityTokenReference.*><wsse:Reference URI="#null"/></wsse:SecurityTokenReference>`;
+        if securedEnvelope.includesMatch(securityToken) {
+            securedEnvelope = regex:replace(securedEnvelope, securityToken.toString(), replace);
+        }
+        else if securedEnvelope.includesMatch(re `<wsse:SecurityTokenReference .*/>`) {
+            securedEnvelope = regex:replace(securedEnvelope, string `<wsse:SecurityTokenReference .*/>`, replace);
+        }
+        return securedEnvelope;
+    } on fail var e {
+        return error Error(e.message());
+    }
+}
+
+function convertStringToXml(string envelope) returns xml|Error {
+    do {
+        return check xml:fromString(regex:replace(envelope, string `<?.*?><`, "<"));
+    } on fail var e {
+        return error Error(e.message());
+    }
```

Comment:
We could use the if checks for error handling cases.

Reply:
Fixed in [1de6750](https://github.com/ballerina-platform/module-ballerina-soap/pull/103/commits/1de6750f3a32d73e90b0ade50476774f2541fccf)